proof-systems/src/darlin/benches/t_dlog_acc_marlin.rs

```diff
@@ -0,0 +1,668 @@
+                }),
+                &a_k_minus_1_plus_a_k_minus_2,
+                &b_k,
+            )?;
+
+            a_k_minus_2 = a_k_minus_1;
+            b_k_minus_2 = b_k_minus_1;
+            a_k_minus_1 = a_k;
+            b_k_minus_1 = b_k;
+        }
+
+        Ok(())
+    }
+}
+
+fn generate_keys<G1: Curve, G2: Curve, CS: ConstraintSynthesizer<G1::ScalarField>>(
+    num_constraints: usize,
+    segment_size: usize,
+    circuit: CS,
+) -> (
+    CommitterKey<G1>,
+    VerifierKey<G1>,
+    ProverKey<G1, G2, D>,
+    MarlinVerifierKey<G1, G2, D>,
+) {
+    let universal_srs =
+        TDLogAccMarlin::<G1, G2, D>::universal_setup(num_constraints, num_constraints, false)
+            .unwrap();
+    let (pc_pk, pc_vk) = universal_srs.trim(segment_size).unwrap();
+    let (index_pk, index_vk) =
+        TDLogAccMarlin::<G1, G2, D>::circuit_specific_setup(&pc_pk, circuit).unwrap();
+    (pc_pk, pc_vk, index_pk, index_vk)
+}
+
+fn generate_trivial_accumulators(
+    pc_pk_g1: &CommitterKey<G1>,
+    pc_vk_g1: &VerifierKey<G1>,
+    index_vk_g1: &MarlinVerifierKey<G1, G2, D>,
+    pc_pk_g2: &CommitterKey<G2>,
+    pc_vk_g2: &VerifierKey<G2>,
+    index_vk_g2: &MarlinVerifierKey<G2, G1, D>,
+) -> (
+    DualDLogItem<G2, G1>,
+    DualSumcheckItem<G2, G1>,
+    DensePolynomial<<G1 as Group>::ScalarField>,
+    DensePolynomial<<G1 as Group>::ScalarField>,
+) {
+    let dlog_acc = DualDLogItem::generate_trivial::<D>(&pc_pk_g2, &pc_pk_g1);
+    let t_acc = DualSumcheckItem::<G2, G1>::generate_trivial();
+
+    // Perform hard verification of the two accumulators in order to compute the
+    // respective polynomials.
+    let (_, verification_g1) = TDLogAccMarlin::<G2, G1, D>::hard_verify(
+        &pc_vk_g2,
+        &pc_vk_g1,
+        &index_vk_g2,
+        &index_vk_g1,
+        &t_acc,
+        &dlog_acc,
+    )
+    .unwrap()
+    .unwrap();
+
+    let bullet_poly = verification_g1.bullet_poly;
+    let t_poly = verification_g1.t_acc_poly;
+
+    (dlog_acc, t_acc, bullet_poly, t_poly)
+}
+
+fn generate_random_accumulators<R: RngCore>(
+    pc_pk_g1: &CommitterKey<G1>,
+    pc_vk_g1: &VerifierKey<G1>,
+    index_pk_g1: &ProverKey<G1, G2, D>,
+    index_vk_g1: &MarlinVerifierKey<G1, G2, D>,
+    pc_pk_g2: &CommitterKey<G2>,
+    pc_vk_g2: &VerifierKey<G2>,
+    index_pk_g2: &ProverKey<G2, G1, D>,
+    index_vk_g2: &MarlinVerifierKey<G2, G1, D>,
+    rng: &mut R,
+) -> (
+    DualDLogItem<G2, G1>,
+    DualSumcheckItem<G2, G1>,
+    DensePolynomial<<G1 as Group>::ScalarField>,
+    DensePolynomial<<G1 as Group>::ScalarField>,
+) {
+    let dlog_acc = DualDLogItem::generate_random::<_, D>(rng, &pc_pk_g2, &pc_pk_g1);
+
+    let t_acc = DualSumcheckItem::<G2, G1>::generate_random::<D>(
+        rng,
+        &index_pk_g2.index_vk.index,
+        &index_pk_g1.index_vk.index,
+        &pc_pk_g2,
+        &pc_pk_g1,
+    );
+
+    // Perform hard verification of the two accumulators in order to compute the
+    // respective polynomials.
```

Comment:
Same comment as before

Reply:
Done, as part of a more general restructuring of accumulators